[fx] Improve support for tuple subclasses such as NamedTuple

jfix71 opened the pull request:
Summary:
Previously, if an arg to an FX node is a subclass of tuple/list/dict then it gets sanitized essentially back to that base class. An example here is when setting an arg to be a TensorMetadata object, which is a NamedTuple, it will be set as a tuple instead.

- Change `map_aggregate` to repack the tuple to `type(a)` when it's not directly a tuple (try/except for best attempt)
- During codegen, call `add_global` for `type(a)` if it's not directly a tuple.
- Add an option for an arg to provide a `_custom_fx_repr_fn` for use inside stringifying via `_format_arg`

Test Plan: Added unit test coverage, where we inline the named tuple into arg/kwarg.

Differential Revision: D34381888

jamesr66a commented on torch/fx/graph.py:
```diff
@@ -357,6 +350,19 @@ def type_repr(o : Any):
 
+        def _format_args(args: Tuple[Argument, ...], kwargs: Dict[str, Argument]) -> str:
+            def _get_repr(arg):
+                if isinstance(arg, tuple) and type(arg) != tuple:
```
I think it's possible that there are tuple subclasses that are not `NamedTuple`. Can we make this more specific by checking if there's a `_fields` attribute: https://stackoverflow.com/a/2166841 (this solution for checking for a `NamedTuple` is ridiculous but it's python's fault lol)

jfix71 replied:
Done.

jfix71 committed:
[fx] Improve support for tuple subclasses such as NamedTuple (#73198)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/73198

Previously, if an arg to an FX node is a subclass of tuple then it gets sanitized essentially back to that base class. An example here is when setting an arg to be a TensorMetadata object, which is a NamedTuple, it will be set as a tuple instead.

- Change `map_aggregate` to repack the tuple to `type(a)` when it's not directly a tuple (try/except for best attempt)
- During codegen, call `add_global` for `type(a)` if it's not directly a tuple.
- Add an option for an arg to provide a `_custom_fx_repr_fn` for use inside stringifying via `_format_arg`

Test Plan: Added unit test coverage, where we inline the named tuple into arg/kwarg.

Differential Revision: D34381888

fbshipit-source-id: 74b93a49e8644a1ca3a184245580d0875a923a3b
